Source/Core/DolphinTool/ExtractCommand.cpp

```diff
@@ -0,0 +1,313 @@
+
+  if (options.is_set("output"))
+    std::filesystem::create_directories(output_folder_path);
+
+  if (gameonly)
+    specific_partition = std::string("data");
+
+  if (const std::unique_ptr<DiscIO::BlobReader> blob_reader =
+          DiscIO::CreateBlobReader(input_file_path);
+      !blob_reader)
+  {
+    fmt::println(std::cerr, "Error: Unable to open disc image");
+    return EXIT_FAILURE;
+  }
+
+  const std::unique_ptr<DiscIO::Volume> disc_volume = DiscIO::CreateVolume(input_file_path);
+
+  if (!disc_volume)
+  {
+    fmt::println(std::cerr, "Error: Unable to open volume");
+    return EXIT_FAILURE;
+  }
+
+  if (!VolumeSupported(*disc_volume))
+    return EXIT_FAILURE;
+
+  if (options.is_set_by_user("list"))
+  {
+    ListVolume(*disc_volume, list_path, specific_partition);
+    return EXIT_SUCCESS;
+  }
+
+  bool extracted_one = false;
+
+  if (disc_volume->GetPartitions().empty())
+  {
+    if (options.is_set("partition"))
+      fmt::println(
+          std::cerr,
+          "Warning: --partition has a value even though this image doesn't have any partitions.");
+
+    if (!options.is_set("single"))
+    {
+      ExtractPartition(*disc_volume, DiscIO::PARTITION_NONE, output_folder_path, mute);
+      extracted_one = true;
+    }
+    else if (auto file_info = GetFileInfo(*disc_volume, DiscIO::PARTITION_NONE, single_file_path);
+             file_info != nullptr)
+    {
+      std::string file;
+      file.append(output_folder_path).append("/");
+
+      file.append(single_file_path);
+
+      if (file_info->IsDirectory())
+        ExtractDirectory(*disc_volume, DiscIO::PARTITION_NONE, single_file_path, output_folder_path,
+                         mute);
+      else
+        ExtractFile(*disc_volume, DiscIO::PARTITION_NONE, single_file_path, file);
+
+      extracted_one = true;
+    }
+  }
+  else
+  {
+    for (DiscIO::Partition& p : disc_volume->GetPartitions())
+    {
+      if (const std::optional<u32> partition_type = disc_volume->GetPartitionType(p))
+      {
+        const std::string partition_name = DiscIO::NameForPartitionType(*partition_type, true);
+
+        if (!specific_partition.empty() &&
+            !Common::CaseInsensitiveEquals(specific_partition, partition_name))
+          continue;
+
+        extracted_one = true;
+
+        std::string file;
+        file.append(output_folder_path).append("/");
+
+        if (!options.is_set("single"))
+        {
+          file.append(partition_name);
+          ExtractPartition(*disc_volume, p, file, mute);
+          continue;
+        }
+
+        if (auto file_info = GetFileInfo(*disc_volume, p, single_file_path); file_info != nullptr)
+        {
+          file.append(single_file_path);
+
+          if (file_info->IsDirectory())
+            ExtractDirectory(*disc_volume, p, single_file_path, output_folder_path, mute);
+          else
+            ExtractFile(*disc_volume, p, single_file_path, file);
+          continue;
+        }
+        extracted_one = false;
```

Comment:
You have hidden a bug here: When you extract a file in a subdirectory, the directory isn't created. So, `extract -i brawl.rvz -o tmp -p DATA -s sora.str` works fine and creates a `tmp/sora.str`, but `extract -i brawl.rvz -o tmp -p DATA -s fighter/captain/FitCaptain.pac` seemingly creates nothing (despite the command claiming to have succeeded). If you manually create the `fighter` and `captain` subdirectories it will extract correctly to `tmp/fighter/captain/FitCaptain.pac`.

So either create the subdirectory here, or just extract the file to `tmp/FitCaptain.pac`. Which of these two is again arguable, but I personally would just extract to `tmp/FitCaptain.pac`.